### DISCUSSION TITLE
UMD centric build and distribution

### DISCUSSION THREAD

**Discussion by jeetiss** [TanStack/query#3986] [General] [answered]
Hello, folks! Thanks for a great library

I have started to use `react-query` and faced that webpack uses UMD build while `package.json` describes a module field that supports tree-shaking. so I dig into build/distribution process and found following problems

## browser field
webpack's resolve module algorithm picks a more specific field from `package.json` (and other bundlers too), so webpack chooses UMD module from `browser` field on the default and that breaks tree-shaking

<details>
    <summary>@tanstack/query-example-react-nextjs</summary>
    
<img width="1506" alt="Снимок экрана 2022-08-05 в 11 12 48" src="https://user-images.githubusercontent.com/6726016/183033639-20b33c71-5c81-442c-a711-baa3b266f0dc.png">

</details>

Why `browser` field is specified in packages? Can I remove it?

I can find any reason for that, readme uses the full path to UMD bundle in CDN example
https://tanstack.com/query/v4/docs/installation#cdn

## `preserveModules` 

CJS build contains `preserveModules` enabled and that creates next problems
- all external modules are copied to the `dist` folder and shipped to `npm`, so all users download them twice and `react-query` has ginormous install size
- `broadcast-channel` have a browser field and `preserveModules` don't copy files specified by this field, so it creates issues like https://github.com/TanStack/query/issues/3965

I can't find any information on why `preserveModules` is here, can I remove it?

@DamianOsipiuk asked about this [here](https://github.com/TanStack/query/pull/3892#issuecomment-1191268011), but nobody answered

## external dependencies specified only for UMD

`react-query` install size becomes big because all external libraries are installed 4 times to `node_modules` ones for each type of distribution (`esm` `cjs` `umd`) and ones because dependence specified in `package.json`

Is it special or I can fix that?

https://packagephobia.com/result?p=%40tanstack%2Freact-query

**Comment by DamianOsipiuk** [comment, marked as answer]
Hi all of those issues should be fixed by https://github.com/TanStack/query/pull/3892, yet it has not been released yet.
We first want to do a Prerelease and give it to community to test how it behaves with different bundlers and setups. And if everything will be ok, we will release it as a normal build.
This should happen soon.
